captum/_utils/sample_gradient.py

```diff
@@ -0,0 +1,176 @@
+    - https://github.com/cybertronai/autograd-hacks
+    - https://github.com/pytorch/opacus/tree/main/opacus/grad_sample
+    """
+
+    def __init__(self, model):
+        self.model = model
+        self.hooks_added = False
+        self.activation_dict = defaultdict(list)
+        self.gradient_dict = defaultdict(list)
+        self.forward_hooks = []
+        self.backward_hooks = []
+
+    def add_hooks(self):
+        self.hooks_added = True
+        self.model.apply(self._register_module_hooks)
+
+    def _register_module_hooks(self, module: torch.nn.Module):
+        if isinstance(module, tuple(SUPPORTED_MODULES.keys())):
+            self.forward_hooks.append(
+                module.register_forward_hook(self._forward_hook_fn)
+            )
+            self.backward_hooks.append(
+                module.register_full_backward_hook(self._backward_hook_fn)
+            )
+
+    def _forward_hook_fn(
+        self,
+        module: Module,
+        module_input: Union[Tensor, Tuple[Tensor, ...]],
+        module_output: Union[Tensor, Tuple[Tensor, ...]],
+    ):
+        inp_tuple = _format_tensor_into_tuples(module_input)
+        self.activation_dict[module].append(inp_tuple[0].clone().detach())
+
+    def _backward_hook_fn(
+        self,
+        module: Module,
+        grad_input: Union[Tensor, Tuple[Tensor, ...]],
+        grad_output: Union[Tensor, Tuple[Tensor, ...]],
+    ):
+        grad_output_tuple = _format_tensor_into_tuples(grad_output)
+        self.gradient_dict[module].append(grad_output_tuple[0].clone().detach())
+
+    def remove_hooks(self):
+        self.hooks_added = False
+
+        for hook in self.forward_hooks:
+            hook.remove()
+
+        for hook in self.backward_hooks:
+            hook.remove()
+
+        self.forward_hooks = []
+        self.backward_hooks = []
+
+    def reset(self):
+        self.activation_dict = defaultdict(list)
+        self.gradient_dict = defaultdict(list)
+
+    def compute_param_sample_gradients(self, loss_blob, loss_mode="mean"):
+        assert (
+            loss_mode.upper() in LossMode.__members__
+        ), f"Provided loss mode {loss_mode} is not valid"
+        mode = LossMode[loss_mode.upper()]
+
+        self.model.zero_grad()
+        loss_blob.backward(gradient=torch.ones_like(loss_blob))
+
+        for module in self.activation_dict:
+            sample_grad_fn = SUPPORTED_MODULES[type(module)]
+            activations = self.activation_dict[module]
+            gradients = self.gradient_dict[module]
+            assert len(activations) == len(gradients), (
+                "Number of saved activations do not match number of saved gradients."
+                " This may occur if multiple forward passes are run without calling"
+                " reset or computing param gradients."
+            )
+            for i, (act, grad) in enumerate(
+                zip(activations, list(reversed(gradients)))
```

Comment:
Reversing grads since when a module is used multiple times, the activations will be aligned with the reverse order of the gradients, since the order is reversed in backprop. Sure, will add a description.